### DISCUSSION TITLE
Add Github Workflow to autoupdate addons

### DISCUSSION THREAD

**Discussion by LucasOe** [LuaLS/LLS-Addons#138] [Ideas]
Currently addons don't get updated automatically. Would it be possible to add a Github Workflow that automatically updates addons to the latest upstream version?

**Comment by carsakiller** [comment]
I think this could lead to a lot of unnecessary updates and PRs. I think what may be a better idea would be publishing an action that can automatically open a new PR on LLS-Addons with your addon. This way, contributors can set up their CI/CD however they like and just send their addon over when it is ready. Let me know what you think.

**Reply by bavalpey** [reply]
The issue with this is that this would add a .github folder inside the add-on. That means that when cloning the submodule, the .github folder gets cloned as well, for no good reason.

**Reply by carsakiller** [reply]
Hmm… that is true. Such are the problems with using Git as a sort of package manager 😄. I'll see how we could go about this. I'm thinking there will be an action that runs every few days that updates all the submodules and then opens PRs for each updated one.

**Reply by bavalpey** [reply]
Well, I am actually quite annoyed with this limitation anyway.

I really want to setup actions for my own Add On, or add in a README, or any number of things. But I can't since that'll propagate downstream.

Also can't do these things in their own branch, since the .gitHub folder at least needs to be in all the branches it interacts with.

I think I like the submodule approach, but could it be smarter somehow? Possible to add excluded folders or something when cloning?

**Reply by bavalpey** [reply]
What if there was a MANIFEST file?

**Reply by carsakiller** [reply]
I don't believe it is possible to ignore files in a submodule when cloning them. Just like a normal Git repo, you get all of what you see. Ideally we would use something like LuaRocks or some custom method made more for this purpose, as discussed in https://github.com/LuaLS/lua-language-server/discussions/2408.

**Reply by luttje** [reply]
My two cents: For the Garry's Mod addon we have [a workflow that publishes the LuaLS-Addon library to a separate branch](https://github.com/luttje/glua-api-snippets/blob/22c11e77ce04c396726bea56f6ee94826f837915/.github/workflows/release.yml#L36-L99).

From there I'm quite certain in the same GH Actions workflow process/chain you could
1. Clone a LuaLS-Addons fork (likely will require a Personal Access Token/PAT, since GITHUB_TOKEN may not have permissions), 
2. Sync it with upstream, 
3. Create an update branch for your addon, 
4. Check that out 
5. Update your submodule
6. Finally then run the [github CLI `pr create` command](https://cli.github.com/manual/gh_pr_create) as [described here](https://stackoverflow.com/a/73340290)

Using the API/CLI and a PAT you can do pretty much anything. Would only require you refreshing the PAT every year since that's the maximum expiry date you can set.

That said we'll keep our PR process manual, since reviewing them is also a manual process. It feels harsh to flood carsakiller with a bunch of automatically created PR's to review.

Hope this helps. Good luck!

**Reply by carsakiller** [reply]
I maintain my own addon, which admittedly requires infrequent updates, but I still see the need for at least semi-automatic PRs. A solution like what @luttje proposes could be ideal. A manually triggered workflow that accomplishes all of those steps should allow for addons to be updated with minimal effort, without sending updates for every line ending change.

I am currently too busy to develop a publishing workflow for use across all addons, but if someone were to make one, I would gladly add it to the README 🙂.
